### PR TITLE
Add `OutputResult` [1/2]

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

**This PR is just moving code around in preparation for #9424**

No logic changes

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -72,6 +72,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
     * Refactored `TrainingBatchLoop` and extracted `OptimizerLoop`, splitting off automatic optimization into its own loop ([#9191](https://github.com/PyTorchLightning/pytorch-lightning/pull/9191))
     * Removed `TrainingBatchLoop.backward()`; manual optimization now calls directly into `Accelerator.backward()` and automatic optimization handles backward in new `OptimizerLoop` ([#9265](https://github.com/PyTorchLightning/pytorch-lightning/pull/9265))
     * Extracted `ManualOptimization` logic from `TrainingBatchLoop` into its own separate loop class ([#9266](https://github.com/PyTorchLightning/pytorch-lightning/pull/9266))
+    * Added `OutputResult` and `Manualresult` classes ([#9437](https://github.com/PyTorchLightning/pytorch-lightning/pull/9437))
```
Comment (awaelchli) [inline CHANGELOG.md:75]:
```suggestion
    * Added `OutputResult` and `ManualResult` classes ([#9437](https://github.com/PyTorchLightning/pytorch-lightning/pull/9437))
```

File: pytorch_lightning/loops/optimization/closure.py [hunk 2]
```diff
@@ -122,69 +56,10 @@ def consume_result(self) -> ClosureResult:
...
-            if self._zero_grad_fn is not None:
-                with self._profiler.profile("zero_grad"):
-                    self._zero_grad_fn()
-
-            if self._backward_fn is not None and step_output.closure_loss is not None:
-                with self._profiler.profile("backward"):
-                    step_output.closure_loss = self._backward_fn(step_output.closure_loss)
-
-        return step_output
+        return self
```
Comment (awaelchli) [inline pytorch_lightning/loops/optimization/closure.py:65]: return self? hmm?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/optimization/closure.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/optimization/closure.py:65]: I chose to return self to chain

`closure().consume_result()`

But it's unused. It can also return None

File: pytorch_lightning/loops/optimization/manual_loop.py [hunk 3]
```diff
@@ -11,16 +11,89 @@
...
+        loss: A detached copy of the closure loss.
+        extra: Any keys other than the loss returned.
+    """
+
+    closure_loss: Optional[Tensor]
+    loss: Optional[Tensor] = field(init=False, default=None)
+    extra: Dict[str, Tensor] = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        # TODO: remove with the deprecation removal in v1.6
```
Comment (tchaton) [inline pytorch_lightning/loops/optimization/manual_loop.py:50]: I am not sure about the `deprecation` there. I believe users won't detach properly the extras as it wasn't done in the past.

File: pytorch_lightning/loops/optimization/manual_loop.py [hunk 4]
```diff
@@ -11,16 +11,89 @@
...
+
+        self._clone_loss()
+
+    def _clone_loss(self) -> None:
+        if self.closure_loss is not None:
+            # the loss will get scaled for amp. avoid any modifications to it
+            self.loss = self.closure_loss.detach().clone()
+
+    @classmethod
+    def from_training_step_output(
```
Comment (tchaton) [inline pytorch_lightning/loops/optimization/manual_loop.py:62]: Can we avoid the code duplication between the Result Objects ?

File: pytorch_lightning/loops/optimization/manual_loop.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loops/optimization/manual_loop.py:50]: Changing this would be out of the scope of this PR.

The plan is to keep the `grad_fn` check but not `detach` it automatically, instead fail.
This was chosen to avoid silent errors where users don't know their outputs have a `grad_fn`.

File: pytorch_lightning/loops/optimization/manual_loop.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/loops/optimization/manual_loop.py:62]: The code duplication is the purpose of this PR haha

[2/2] will change the implementations, I split it up so it's easier to review

Because if you move and change the code in the same PR, then it's very hard to review as you lose the side-by-side diff

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by kaushikb11 — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]
not very familiar with the code, but LGTM
